File: cpp/leetcode/200.cpp
class Solution {
public:
    int numIslands(vector<vector<char>> &grid) {
        if (grid.empty() || grid[0].empty()) {
            return 0;
        }
        int m = grid.size(), n = grid[0].size();
        vector<vector<bool>> visit(m, (vector<bool>(n, false)));
        int cnts = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1' && !visit[i][j]) {
                    cnts++;
                    dfs(i, j, m, n, grid, visit);
                }
            }
        }
        return cnts;
    }

    void dfs(int i, int j, int m, int n, vector<vector<char>> &grid, vector<vector<bool>> &visit) {
        if (i < 0 || j < 0 || i >= m || j >= n || grid[i][j] == '0' || visit[i][j]) {
            return;
        }
        visit[i][j] = true;
        dfs(i + 1, j, m, n, grid, visit);
        dfs(i - 1, j, m, n, grid, visit);
        dfs(i, j + 1, m, n, grid, visit);
        dfs(i, j - 1, m, n, grid, visit);
    }
};

class Solution {
public:
    int numIslands(vector<vector<char>> &grid) {
        if (grid.empty()) {
            return 0;
        }
        int m = grid.size(), n = grid[0].size();
        int cnts = 0;
        vector<vector<bool>> check(m, vector<bool>(n, false));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1' && !check[i][j]) {
                    bfs(i, j, m, n, check, grid);
                    cnts++;
                }
            }
        }
        return cnts;
    }

    void bfs(int i, int j, int m, int n, vector<vector<bool>> &check, vector<vector<char>> &grid) {
        check[i][j] = true;
        queue<vector<int>> q;
        q.push({i, j});
        while (!q.empty()) {
            int x = q.front()[0], y = q.front()[1];
            q.pop();
            if (x - 1 >= 0 && !check[x - 1][y] && grid[x - 1][y] == '1') {
                q.push({x - 1, y});
                check[x - 1][y] = true;
            }
            if (x + 1 < m && !check[x + 1][y] && grid[x + 1][y] == '1') {
                q.push({x + 1, y});
                check[x + 1][y] = true;
            }
            if (y - 1 >= 0 && !check[x][y - 1] && grid[x][y - 1] == '1') {
                q.push({x, y - 1});
                check[x][y - 1] = true;
            }
            if (y + 1 < n && !check[x][y + 1] && grid[x][y + 1] == '1') {
                q.push({x, y + 1});
                check[x][y + 1] = true;
            }
        }
    }
};

File: python/leetcode/200.py
class Solution:
    def numIslands(self, grid):
        ans = 0
        if not len(grid):
            return ans
        m, n = len(grid), len(grid[0])
        visited = [[False] * n for i in range(m)]
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1' and not visited[i][j]:
                    ans += 1
                    self.bfs(grid, visited, i, j, m, n)
        return ans

    def bfs(self, grid, visited, i, j, m, n):
        dz = zip([1, 0, -1, 0], [0, 1, 0, -1])
        queue = [(i, j)]
        visited[i][j] = True
        while queue:
            front = queue.pop(0)
            for p in dz:
                np = (front[0] + p[0], front[1] + p[1])
                if self.isValid(np, m, n) and grid[np[0]][np[1]] == '1' and not visited[np[0]][np[1]]:
                    visited[np[0]][np[1]] = True
                    queue.append(np)

    def isValid(self, np, m, n):
        return 0 <= np[0] < m and 0 <= np[1] < n


class Solution(object):
    def numIslands(self, grid):
        ans = 0
        m, n = len(grid), len(grid[0])
        check = [[False for i in xrange(n)] for j in xrange(m)]
        for i in xrange(m):
            for j in xrange(n):
                if not check[i][j] and grid[i][j] == '1':
                    self.dfs(grid, check, i, j, m, n)
                    ans += 1
        return ans

    def dfs(self, grid, check, i, j, m, n):
        check[i][j] = True
        if i > 0 and not check[i - 1][j] and grid[i - 1][j] == '1':
            self.dfs(grid, check, i - 1, j, m, n)
        if i < m - 1 and not check[i + 1][j] and grid[i + 1][j] == '1':
            self.dfs(grid, check, i + 1, j, m, n)
        if j > 0 and not check[i][j - 1] and grid[i][j - 1] == '1':
            self.dfs(grid, check, i, j - 1, m, n)
        if j < n - 1 and not check[i][j + 1] and grid[i][j + 1] == '1':
            self.dfs(grid, check, i, j + 1, m, n)
